torch/_subclasses/functional_tensor.py

```diff
@@ -0,0 +1,128 @@
+        #   which is of type FunctionalTensorWrapper! We explicitly do not want our wrapper to be a FunctionalTensorWrapper.
+        # - If we use the default tensor.__new__(), we have another problem: it returns inner_tensor.alias(),
+        #   which causes every subclass created above autograd to have autograd view metadata (in addition to also being a FunctionalTensorWrapper).
+        raise RuntimeError(
+            "Attempting to use FunctionalTensor on its own. Instead, please use it with a corresponding FunctionalTensorMode()"
+        )
+
+    @staticmethod
+    def to_functional(x):
+        # We will do the wrapping for the user.
+        assert not torch._is_functional_tensor(x)
+        # The only autograd metadata we care about on the FunctionalTensor is:
+        # - requires_grad (so autograd runs)
+        # - is_leaf (so that mutations on graph inputs that are not leaves are allowed by the autograd engine)
+        #   this is handled by FunctionalTensor.to_functional
+        x_functional = torch._to_functional_tensor(x, mirror_autograd_meta=True)
+
+        torch._functionalize_enable_reapply_views(True)
+        if x.requires_grad and not x.is_leaf:
+            out = _ToFunctionalTensor.apply(x_functional)
+            return out
+        else:
+            out = FunctionalTensor(x_functional)
+            out.requires_grad = x_functional.requires_grad
+            return out
+
+
+class FunctionalTensorMode(TorchDispatchMode):
+    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+        if kwargs is None:
+            kwargs = {}
+
+        def assert_is_functional(x):
+            assert torch._is_functional_tensor(x)
+
+        def wrap(x):
+            return FunctionalTensor(x)
+
+        def unwrap(x):
+            return x.elem
+
+        # Expectation: functionalization should not **already** be enabled above our mode.
+        # Why would that be bad? when we return a FunctionalTensor here, we don't want functionalization
+        # to run above this mode and further wrap that output in **another** C++ FunctionalTensorWrapper.
+        is_included = torch._C._dispatch_tls_is_dispatch_key_included(
+            torch._C.DispatchKey.Functionalize
+        )
+        is_excluded = torch._C._dispatch_tls_is_dispatch_key_excluded(
+            torch._C.DispatchKey.Functionalize
+        )
+        assert is_excluded or not is_included
+        # All we want to do here is re-use the existing C++ functionalization logic.
+        # This requires swizzling our TLS dispatch keys so that the Functionalize key is active.
+        with torch._C._SetExcludeDispatchKeyGuard(
+            torch._C.DispatchKey.Functionalize, False
+        ):
+            try:
+                # By default for python functionalization (for AOTAutograd), we reapply views.
+                torch._functionalize_enable_reapply_views(True)
```

Comment:
so this no-ops if functionalization was already included, right?

Reply:
unless TLS reentrant rears its head, which I can't easily tell 🤔

Reply:
Mentioned below - this doesn't actually toggle the `Functionalize` dispatch key TLS.

But I am in the `IncludeDispatchKeyGuard` context manager above (which will indeed nop if it was already included).